### PR TITLE
Process ingestions and gather data from GitHub

### PR DESCRIPTION
**Pull request by rowanmanning**
Now the ingestion queue processor gathers data from GitHub and creates
Version entities from Ingestion entities. There's a lot of code here,
but it wasn't really possible to split much more than this and have
something functional in place.

If everything goes well, an Ingestion will be converted into a Version
when all of the available information is fetched from GitHub. If an
error occurs:

  1. We determine whether the error is recoverable or not. An example of
    a recoverable error is if the rate limit has been exceeded. An
    example of a non-recoverable error is if the repo/tag does not
    include and Origami manifest.

    1. For recoverable errors we increment the number of attempts and will
       try again later.

    2. For non-recoverable errors we remove the ingestion and do not
       attempt again.

I've also added logic to prevent an Ingestion from being created if it
already exists or if it's already been converted to a Version. This
reduces our GitHub API usage.

Running locally this has ingested 500-600 versions per hour, it would be
faster if we didn't keep hitting the GitHub API rate limit. I'm not
going to address this as we're highly unlikely to have this many
releases on GitHub in any given hour.

I haven't added in cleanup of over-attempted or long-running ingestions
yet, this will come in a separate PR. I've also not found a nice way to
run integration tests on this part of the code-base, in the interest of
brevity for this already-large set of changes, I'll add tests in a
separate PR as well.

### CODE REVIEW

File: data/migration/20171124163526_remove-commit-hash.js [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+	// Remove the commit_hash column
+	await database.schema.table('versions', table => {
+		table.dropColumn('commit_hash');
+	});
+
+};
+
+exports.down = async () => {
+
+	// We can't recover from this 😱
```
Comment (JakeChampion) [inline data/migration/20171124163526_remove-commit-hash.js]: 🙀

Comment (rowanmanning) [reply data/migration/20171124163526_remove-commit-hash.js]: Haha in this case it's OK, there was never any data stored in the `commit_hash` field on any running instances of the app except mine. I'm tempted, before going into production, to squash all the migrations down into one file

File: lib/github-client.js [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+		if (githubAuthToken) {
+			this.client.authenticate({
+				type: 'oauth',
+				token: githubAuthToken
+			});
+		}
+	}
+
+	extractRepoFromUrl(url) {
+		const [, , owner, repo] = url.match(module.exports.githubUrlRegexp);
```
Comment (JakeChampion) [inline lib/github-client.js]: Fancy :-)

File: lib/routes/v1/queue.js [hunk 3]
```diff
@@ -78,7 +78,8 @@ module.exports = app => {
 
 		} catch (error) {
 			if (error.name === 'ValidationError') {
-				return response.status(400).render('error', {
+				const status = (error.isConflict ? 409 : 400);
+				return response.status(status).render('error', {
 					title: 'Error 400',
```
Comment (JakeChampion) [inline lib/routes/v1/queue.js]: should the title change from 400 to 409 if the status is a 409?

File: models/version.js [hunk 4]
```diff
@@ -166,25 +180,173 @@ function initModel(app) {
...
+				const tag = ingestion.get('tag');
 
-			// TODO get the full information from GitHub
-			const version = new Version({
-				name: ingestion.get('url').replace(/^https?:\/\/(www\.)?github.com\/|\/$/gi, ''),
-				url: ingestion.get('url'),
-				tag: ingestion.get('tag'),
-				commit_hash: 'TODO'
-			});
+				// Expect a valid GitHub URL
```
Comment (JakeChampion) [inline models/version.js]: Note: I don't think we can switch to this new service until after 21st Feb, that is the date when Stash will be turned off and all repositories will be on GitHub.

File: lib/routes/v1/queue.js (hunk 3 above)
Comment (rowanmanning) [reply lib/routes/v1/queue.js]: Yep, well spotted. I've changed

**Pull request merged**